src/dom.js

```diff
@@ -561,23 +560,23 @@ export function getDataParamsFromAttributes(element, opt_computeParamNameFunc,
 }
 
 /**
- * Whether the element have a next node in the document order.
+ * Whether the Node has a next Node in the document order.
  * This means either:
- *  a. The element itself has a nextSibling.
- *  b. Any of the element ancestors has a nextSibling.
- * @param {!Element} element
- * @param {?Node} opt_stopNode
+ *  a. The Node itself has a nextSibling.
+ *  b. Any of the Node ancestors has a nextSibling.
+ * @param {!Node} currentNode The Node to look for a following Node.
+ * @param {?Node} root A root Node to look within.
+ * @param {?NodeFilter} filter A filter to exclude certain Nodes from being
+ *    considered as a following Node.
  * @return {boolean}
  */
-export function hasNextNodeInDocumentOrder(element, opt_stopNode) {
-  let currentElement = element;
-  do {
-    if (currentElement.nextSibling) {
-      return true;
-    }
-  } while ((currentElement = currentElement.parentNode) &&
-            currentElement != opt_stopNode);
-  return false;
+export function hasNextNodeInDocumentOrder(
+  currentNode, root = rootNodeFor(currentNode), filter = null) {
+  // We need the 4th argument for IE.
+  const walker = currentNode.ownerDocument.createTreeWalker(
```

Comment:
Tree walkers are **terribly slow**. We can't use this in a hotpath.

Reply:
I'm sure it is slow when actually walking a tree (due to jumping back and forth between C++ and JS for the filter function), but we really shouldn't be jumping back and forth unless you actually hit dynamically appended nodes. This should be extremely rare. You should almost always 999/1000 times simply walk the tree in C++, then confirm for one node that it is as expected.

I'll look into doing it without `TreeWalker` though.

Reply:
They're slow because they're defined by WebIDL, which is an abstraction on top of C++. Just startup time alone, without any special features, it is pretty bad. http://jsbench.github.io/#51e8c9659d33f1c6312b504e9e9e2bbf